### PR TITLE
Changed FID to be online instead of storing each feature

### LINKED ISSUE #1198: Change FID to be online
## 🚀 Feature

The idea is to compute mean and covariance online after each update instead of storing the features and then computing the mean and covariance

### Motivation

The motivation is to avoid storing the features which will lead to a way more memory efficient metric.

### Pitch

This can be done the following way:
```python
def update(self, imgs: Tensor, real: bool) -> None:  # type: ignore
        """Update the state with extracted features.

        Args:
            imgs: tensor with images feed to the feature extractor
            real: bool indicating if ``imgs`` belong to the real or the fake distribution
        """
        features = self.inception(imgs).double()
        self.orig_dtype = features.dtype

        if real:
            self.real_features_sum += features.sum(dim=0)
            self.real_features_cov_sum += features.t().mm(features)
            self.real_features_num_samples += imgs.shape[0]
        else:
            self.fake_features_sum += features.sum(dim=0)
            self.fake_features_cov_sum += features.t().mm(features)
            self.fake_features_num_samples += imgs.shape[0]

    def compute(self) -> Tensor:
        """Calculate FID score based on accumulated extracted features from the two distributions."""
        mean_real = (self.real_features_sum / self.real_features_num_samples).unsqueeze(0)
        mean_fake = (self.fake_features_sum / self.fake_features_num_samples).unsqueeze(dim=0)

        cov_real = (self.real_features_cov_sum / self.real_features_num_samples - mean_real.t().mm(mean_real)) / (
            self.real_features_num_samples - 1
        )
        cov_fake = (self.fake_features_cov_sum - self.fake_features_num_samples * mean_fake.t().mm(mean_fake)) / (
            self.fake_features_num_samples - 1
        )
        return _compute_fid(mean_real, cov_real, mean_fake, cov_fake).to(self.orig_dtype)
```

### PR DESCRIPTION
**Pull request by nicolas-dufour**
## What does this PR do?

Fixes #1198

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: src/torchmetrics/image/fid.py [hunk 1]
```diff
@@ -240,8 +244,13 @@ def __init__(
...
 
-        self.add_state("real_features", [], dist_reduce_fx=None)
-        self.add_state("fake_features", [], dist_reduce_fx=None)
+        self.add_state("real_features_sum", [], dist_reduce_fx=None)
+        self.add_state("real_features_cov_sum", [], dist_reduce_fx=None)
+        self.add_state("real_features_num_samples", torch.tensor(0).long(), dist_reduce_fx="sum")
+
+        self.add_state("fake_features_sum", [], dist_reduce_fx=None)
+        self.add_state("fake_features_cov_sum", [], dist_reduce_fx=None)
+        self.add_state("fake_features_num_samples", torch.tensor(0).long(), dist_reduce_fx="sum")
```
[the remark this answers is not in the document]
Comment (nicolas-dufour) [reply src/torchmetrics/image/fid.py:253]: Yep you are right

File: src/torchmetrics/image/fid.py [hunk 2]
```diff
@@ -201,8 +200,13 @@ class FrechetInceptionDistance(Metric):
...
 
-    real_features: List[Tensor]
-    fake_features: List[Tensor]
+    real_features_sum: List[Tensor]
+    real_features_cov_sum: List[Tensor]
+    real_features_num_samples: int
+
+    fake_features_sum: List[Tensor]
+    fake_features_cov_sum: List[Tensor]
+    fake_features_num_samples: int
```
[the remark this answers is not in the document]
Comment (nicolas-dufour) [reply src/torchmetrics/image/fid.py]: Thanks for the comment, i have made the changes and debug a few bugs. Also had to adapt the tests

File: src/torchmetrics/image/fid.py [hunk 3]
```diff
@@ -219,6 +223,7 @@ def __init__(
         )
 
         if isinstance(feature, int):
+            num_features = int(feature)
```
Comment (SkafteNicki) [inline src/torchmetrics/image/fid.py:226]: should already be int
```suggestion
            num_features = feature
```

File: src/torchmetrics/image/fid.py [hunk 4]
```diff
@@ -250,40 +267,42 @@ def update(self, imgs: Tensor, real: bool) -> None:  # type: ignore
...
+        )
+        return _compute_fid(mean_real.squeeze(0), cov_real, mean_fake.squeeze(0), cov_fake).to(self.orig_dtype)
 
     def reset(self) -> None:
         if not self.reset_real_features:
             # remove temporarily to avoid resetting
-            value = self._defaults.pop("real_features")
+            real_features_sum = self._defaults.pop("real_features_sum")
+            real_features_cov_sum = self._defaults.pop("real_features_cov_sum")
+            real_features_num_samples = self._defaults.pop("real_features_num_samples")
```
Comment (SkafteNicki) [inline src/torchmetrics/image/fid.py:302]: these are just tensors now, so we should just copy the current value -> reset -> restore value
```suggestion
            real_features_sum = deepcopy(self.real_features_sum)
            real_features_cov_sum = deepcopy(self.real_features_cov_sum)
            real_features_num_samples = deepcopy(self.real_features_num_samples)
```

File: src/torchmetrics/image/fid.py [hunk 5]
```diff
@@ -250,40 +267,42 @@ def update(self, imgs: Tensor, real: bool) -> None:  # type: ignore
...
             super().reset()
-            self._defaults["real_features"] = value
+            self._defaults["real_features_sum"] = real_features_sum
+            self._defaults["real_features_cov_sum"] = real_features_cov_sum
+            self._defaults["real_features_num_samples"] = real_features_num_samples
```
Comment (SkafteNicki) [inline src/torchmetrics/image/fid.py:306]:
```suggestion
            self.real_features_sum = real_features_sum
            self.real_features_cov_sum = real_features_cov_sum
            self.real_features_num_samples = real_features_num_samples
```

**Review by SkafteNicki — APPROVED** [review]
LGTM :)

**Review by Borda — APPROVED** [review]

**Comment by zhou13** [comment]
Hi @nicolas-dufour, could you confirm that 
```python
            self.fake_features_cov_sum += features.t().mm(features)
```
in this PR is the correct way to compute the covariance matrix?

I think this will generate different results when you try to batch the input to save memory, e.g., 
```
fid.update(images[:200]);
fid.compute(...);
```
vs
```
fid.update(images[:100]);
fid.update(images[100:200]);
fid.compute(...);
```
